### PR TITLE
Add StatefulSetGrid workload

### LINKED ISSUE #26: SEP: ServiceGroup StatefulSetGrid Design Specification
This is my initial design of ServiceGroup StatefulSetGrid as below:

image

StatefulSetGrid custom resource samples below:

```yaml
apiVersion: superedge.io/v1
kind: StatefulSetGrid
metadata:
  name: statefulsetgrid-demo
  namespace: default
spec:
  gridUniqKey: zone1
  template:
    selector:
      matchLabels:
        appGrid: nginx # has to match .spec.template.metadata.labels
    serviceName: "servicegrid-demo-svc"
    replicas: 3 # by default is 1
    template:
      metadata:
        labels:
          appGrid: nginx # has to match .spec.selector.matchLabels
      spec:
        terminationGracePeriodSeconds: 10
        containers:
        - name: nginx
          image: k8s.gcr.io/nginx-slim:0.8
          ports:
          - containerPort: 80
            name: web
          volumeMounts:
          - name: www
            mountPath: /usr/share/nginx/html
    volumeClaimTemplates:
    - metadata:
        name: www
      spec:
        accessModes: [ "ReadWriteOnce" ]
        storageClassName: "my-storage-class"
        resources:
          requests:
            storage: 1Gi
```

ServiceGrid stays unchanged and samples below:

```yaml
apiVersion: superedge.io/v1
kind: ServiceGrid
metadata:
  name: servicegrid-demo
  namespace: default
spec:
  gridUniqKey: zone1
  template:
    selector:
      appGrid: nginx
    ports:
    - port: 80
      name: web
    clusterIP: None                                                       
```

this design works as follows:
1. StatefulSetGrid will generate relevant Kubernetes statefulset workload for each of nodeunit as DeploymentGrid does(i.e. {StatefulSetGrid}-{NodeUnit})
2. StatefulSet headless service will be access by domains constructed by `{StatefulSetGrid}-{0..N-1}.{ServiceGrid}-svc.
default.svc.cluster.local` which corresponds to actual statefulset workload `{StatefulSetGrid}-{NodeUnit}-{0..N-1}.{ServiceGrid}-svc.default.svc.cluster.local`, aiming to block the difference of nodeunits. 
3. StatefulSet normal clusterIP service access stays unchanged by using achieved service topology.(i.e. {ServiceGrid}-svc)

### PR DESCRIPTION
**Pull request by duyanghao**
Close #26 
This PR add StatefulSetGrid workload relevant controllers on edge and refactors application-grid-controller as a whole.
Signed-off-by: duyanghao <1294057873@qq.com>

### CODE REVIEW

**Comment by duyanghao** [comment]
@chenkaiyue PTAL.

File: pkg/application-grid-controller/controller/service/util/service_util.go [hunk 1]
```diff
@@ -60,16 +71,27 @@ func constructServicePortIdentify(port *corev1.ServicePort) string {
...
-		copyObj.Labels = make(map[string]string)
+	// Append existed ServiceGrid labels to service to be checked
+	if sg.Labels != nil {
+		copyObj.Labels = sg.Labels
+		copyObj.Labels[common.GridSelectorName] = sg.Name
+		copyObj.Labels[common.GridSelectorUniqKeyName] = sg.Spec.GridUniqKey
+	} else {
+		copyObj.Labels = map[string]string{
+			common.GridSelectorName:        sg.Name,
+			common.GridSelectorUniqKeyName: sg.Spec.GridUniqKey,
```
Comment (chenkaiyue) [inline pkg/application-grid-controller/controller/service/util/service_util.go:86]: Now deployment, statefulset, service all have GridSelectorUniqKeyName to display uniqkey, maybe we can make it as a convention; While different with others ,service have TopologyAnnotationsKey in Annotations before, it have the same effect with GridSelectorUniqKeyName, we can remove it.

Comment (duyanghao) [reply pkg/application-grid-controller/controller/service/util/service_util.go:86]: I recommend to remove it when application-grid-wrapper develops more maturity or deprecated in future.

File: pkg/application-grid-controller/controller/deployment/node.go [hunk 2]
```diff
@@ -32,31 +32,36 @@ func (dgc *DeploymentGridController) addNode(obj interface{}) {
...
+	if curNode.ResourceVersion == oldNode.ResourceVersion {
+		// Periodic resync will send update events for all known Nodes.
+		// Two different versions of the same Node will always have different RVs.
+		return
+	}
 	labelChanged := !reflect.DeepEqual(curNode.Labels, oldNode.Labels)
 	// Only handles nodes whose label has changed.
 	if labelChanged {
-		dgs := dgc.getGridForNode(curNode)
+		dgs := dgc.getGridForNode(oldNode, curNode)
```
Comment (chenkaiyue) [inline pkg/application-grid-controller/controller/deployment/node.go:59]: GetGridForNode -> GetNodesSelector  use GridSelectorUniqKeyName as key, seems can't get wanted deployment grid?

Comment (duyanghao) [reply pkg/application-grid-controller/controller/deployment/node.go:59]: That's right, we need fix it.

**Pull request closed without merging**